File: daily_problem/2023-3/2488.cpp
/*
2488. 统计中位数为 K 的子数组
https://leetcode.cn/problems/count-subarrays-with-median-k/description/
*/

#include <algorithm>
#include <unordered_map>
#include <vector>

using std::find;
using std::unordered_map;
using std::vector;

class Solution {
public:
  int countSubarrays(vector<int> &nums, int k) {
    int p = find(nums.begin(), nums.end(), k) - nums.begin();
    unordered_map<int, int> cnt{{0, 1}};
    int x = 0;
    for (int i = p - 1; i >= 0; --i) {
      x += nums[i] < k ? 1 : -1;
      cnt[x]++;
    }
    int res = cnt[0] + cnt[-1];
    x = 0;
    for (int i = p + 1; i < nums.size(); ++i) {
      x += nums[i] > k ? 1 : -1;
      res += cnt[x] + cnt[x - 1];
    }
    return res;
  }
};

File: daily_problem/2023-3/2488.py
"""
2488. 统计中位数为 K 的子数组
https://leetcode.cn/problems/count-subarrays-with-median-k/description/
"""


class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        p = nums.index(k)
        cnt = Counter({0: 1})
        c = 0
        for i in range(p-1, -1, -1):
            x = nums[i]
            c += 1 if x < k else -1
            cnt[c] += 1
        res = cnt[0] + cnt[-1]
        c = 0
        for x in nums[p+1:]:
            c += 1 if x > k else -1
            res += cnt[c] + cnt[c-1]
        return res
